src/boot/mlang.ml

```diff
@@ -8,140 +8,171 @@
+  let merge_syn _ a b = match a, b with
+    | None, None -> None
+    | None, Some _ -> Some (NoInfo, [])
+    | Some a, None -> Some a
+    | Some (fi, cons), Some (_, old_cons) ->
+       Some (fi, List.filter (fun c1 -> List.exists (eq_cons c1) old_cons |> not) cons) in
+  let merge_inter name a b = match a, b with
+    | None, None -> None
+  | None, Some a -> Some a
+    | Some a, None -> Some a
+    | Some ({info=fi1; params=p1; cases=c1; subsets=s1} as data)
+    , Some {info=fi2; params=p2; cases=c2; subsets=s2} ->
+       if fi1 = fi2 then Some data
+       else if List.length p1 <> List.length p2 then
+         raise_error fi1 ("Different number of parameters for interpreter '" ^ Ustring.to_utf8 name ^ "' compared to previous definitian at " ^ Ustring.to_utf8 (info2str fi2))
+       else
+         let c2 = List.filter (fun (fi, _) -> List.exists (fun (fi2, _) -> eq_info fi fi2) c1 |> not) c2 in
+         let subsets = add_new_by (fun (a1, a2) (b1, b2) -> eq_info a1 b1 && eq_info a2 b2) s2 s1 in
+         let subsets =
+           liftA2 (fun a b -> (a, b)) c1 c2
+           |> fold_map ~fold:(fun a b -> List.rev_append b a) ~map:(compute_order fi1) subsets in
+         Some {data with subsets; cases = List.rev_append c2 c1} in
+  {inters = Record.merge merge_inter i1 i2; syns = Record.merge merge_syn s1 s2}
+
+(* TODO(vipa): this is likely fairly low hanging fruit when it comes to optimization *)
+let topo_sort (eq: 'a -> 'a -> bool) (edges: ('a * 'a) list) (vertices: 'a list) =
+  let rec recur rev_order edges vertices =
+    match List.find_opt (fun v -> List.exists (fun (_, t) -> eq v t) edges |> not) vertices with
+    | None -> List.rev rev_order
+    | Some x ->
+       recur
+         (x :: rev_order)
+         (List.filter (fun (f, t) -> not (eq x f) && not (eq x t)) edges)
+         (List.filter (fun v -> not (eq x v)) vertices)
+  in recur [] edges vertices
+
+let data_to_lang info name includes {inters; syns}: mlang =
+  let info_assoc fi l = List.find (fun (fi2, _) -> eq_info fi fi2) l |> snd in
+  let syns =
+    Record.bindings syns
+    |> List.map (fun (syn_name, (fi, cons)) -> Data(fi, syn_name, cons)) in
+  let sort_inter name {info; params; cases; subsets} =
+    let mk_case fi = let case = info_assoc fi cases in (case.pat, case.rhs) in
+    let cases =
+      List.map fst cases
+      |> topo_sort eq_info subsets
+      |> List.map mk_case in
+    Inter(info, name, params, cases) in
+  let inters =
+    Record.bindings inters
+    |> List.map (fun (name, inter_data) -> sort_inter name inter_data) in
+  Lang(info, name, includes, List.rev_append syns inters)
+
+let flatten_lang (prev_langs: lang_data Record.t): top -> lang_data Record.t * top = function
+  | (TopLet _ | TopRecLet _ | TopCon _) as top -> (prev_langs, top)
+  | TopLang(Lang(info, name, includes, _) as lang) ->
+     let self_data = compute_lang_data lang in
+     let included_data = List.map (lookup_lang info prev_langs) includes in
```

Comment:
This isn't really a list of data, but a list of langs, right?

Reply:
It's a `lang_data list`. I don't particularly like the name of this type, but I didn't come up with anything better. It's supposed to be a representation of the results of the calculations I need for a language for my analyses here, so essentially caching (though I prefer to think of it as a more accurate representation, e.g., the cases are put in a DAG instead of just in a list, since the order in the list doesn't actually matter).